swupd-extract/state.go

```diff
@@ -0,0 +1,367 @@
+			fmt.Printf("- Resetting state in %s previously used for %q\n", stateDir, stateContent)
+			// Delete files individually in case stateDir is managed by the user.
+			var fis []os.FileInfo
+			fis, err = ioutil.ReadDir(stateDir)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't reset state directory: %s", err)
+			}
+			for _, fi := range fis {
+				err = os.RemoveAll(filepath.Join(stateDir, fi.Name()))
+				if err != nil {
+					return nil, fmt.Errorf("couldn't reset state directory: %s", err)
+				}
+			}
+		}
+	}
+	err = os.MkdirAll(filepath.Join(stateDir, "staged/temp"), 0755)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create state directory: %s", err)
+	}
+	err = ioutil.WriteFile(contentFilename, []byte(baseContent), 0644)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't write %s: %s", contentFilename, err)
+	}
+
+	var isRemote bool
+	if strings.HasPrefix(baseContent, "https://") || strings.HasPrefix(baseContent, "http://") {
+		isRemote = true
+	}
+
+	cs := &clientState{
+		dir:         stateDir,
+		baseContent: baseContent,
+		isRemote:    isRemote,
+	}
+
+	return cs, nil
+}
+
+func (cs *clientState) OpenFile(elem ...string) (io.ReadCloser, error) {
+	joined := filepath.Join(elem...)
+	if !cs.isRemote {
+		return os.Open(filepath.Join(cs.baseContent, joined))
+	}
+
+	u := cs.baseContent + "/" + joined
+	fmt.Printf("- downloading %s\n", u)
+	res, err := http.Get(u)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		_ = res.Body.Close()
+		return nil, fmt.Errorf("couldn't download %q: got response with code: %d %s", u, res.StatusCode, http.StatusText(res.StatusCode))
+	}
+	return res.Body, nil
+}
+
+func (cs *clientState) GetFile(elem ...string) (string, error) {
+	joined := filepath.Join(elem...)
+	if !cs.isRemote {
+		return filepath.Join(cs.baseContent, joined), nil
+	}
+	localFile := filepath.Join(cs.dir, joined)
+	if _, err := os.Stat(localFile); err == nil {
+		if !cs.noCache {
+			return localFile, nil
+		}
+		err = os.RemoveAll(localFile)
+		if err != nil {
+			return "", fmt.Errorf("couldn't remove %s to redownload: %s", localFile, err)
+		}
+	}
+	err := os.MkdirAll(filepath.Dir(localFile), 0755)
+	if err != nil {
+		return "", err
+	}
+	err = download(cs.baseContent+"/"+joined, localFile)
+	if err != nil {
+		return "", err
+	}
+	return localFile, nil
+}
+
+func (cs *clientState) Path(elem ...string) string {
+	return filepath.Join(cs.dir, filepath.Join(elem...))
+}
+
+func (cs *clientState) GetBundleManifest(version, name, expectedHash string) (*swupd.Manifest, error) {
+	if name == "MoM" {
+		return nil, fmt.Errorf("invalid arguments to GetBundleManifest: MoM is not a bundle")
+	}
+	filename, err := cs.GetFile(version, "Manifest."+name)
+	if err != nil {
+		return nil, err
+	}
+	hash, err := swupd.GetHashForFile(filename)
+	if err != nil {
+		_ = os.Rename(filename, filename+".invalid")
+		return nil, fmt.Errorf("couldn't calculate hash for %s: %s", filename, err)
```

Comment:
Some indication that the bad file is left in $filename.invalid ,at least of os.Rename doesn't report an error?

Reply:
Makes sense to warn the user.

I'll see if I can also change this to download the file elsewhere and only move it to the final place when happy. That at least diminishes the impact of os.Rename failing.

Reply:
> I'll see if I can also change this to download the file elsewhere and only move it to the final place when happy. That at least diminishes the impact of os.Rename failing.

Nevermind my comment. For this particular case, we are downloading to replicate what is externally AND always verifying the file. I'll remove the .invalid rename, as it's the wrong thing to do if we are dealing with local paths (where the file from the local path repository is used directly).

Reply:
Removed the os.Rename. If the file is bad, we'll keep telling the user about it.